### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by davidlattimore** [google/evcxr#1]
In the unlikely event that it already works on Windows, please let me know.

If you'd like to work on this, feel free to reach out with any questions. I'm not really sure what will be involved, since I don't know what the problems will be.

**Comment by mrobakowski** [comment]
The only dependency that just straight does not compile is `sig`. I managed to make some test pass by stubbing the `install_crash_handlers` in the runtime. There's also some problem with loading dlls, which seems to be solved by prepending the `deps` folder to the `PATH` env variable right before calling `libloading::Library::new`. I think it doesn't find the `evcxr_internal_runtime.dll` without that. I also have some weird problems with access control to the produced files in the tmp folder (ex. linker cannot access the dlls produced by the compiler sometimes), but that may just be a problem with my system.

**Comment by davidlattimore** [comment]
Thanks for looking into that. That's interesting that it needs PATH to be set in order to find the dlls. When compiling I pass -C rpath to rustc, which is supposed to include absolute paths to dependent shared objects. Perhaps that flag doesn't do anything on Windows and it's always necessary to set PATH.

I mentioned this on the Mac porting bug, but... one bit of code that comes to mind with the potential to cause problems is the "pre-warming" code, which after compiling a crate, builds the next crate with the same code to warm up cargo/rustc's caches. From my experiments, this reduced latency from I think ~400ms to ~300ms. You could try disabling that and see if it makes a difference. The line that sets it is [here](https://github.com/google/evcxr/blob/master/evcxr/src/eval_context.rs#L110).

**Comment by mrobakowski** [comment]
Yeah, sadly there's no `rpath` equivalent for windows shared libraries. After a quick search, I found out that [`SetDllDirectoryW`](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-setdlldirectoryw) can be used instead of setting PATH.

**Comment by hajsf** [comment]
I got the below error while installing at 'Windows 10'.
As a work around, I installed [WSL](https://www.microsoft.com/en-ca/p/ubuntu/9nblggh4msv6?rtc=1&activetab=pivot%3Aoverviewtab) and worked as Linux, but if you need me to test/change something at Windows I'm redy to help.

<img width="766" alt="vxr" src="https://user-images.githubusercontent.com/26045838/47899475-b5b21480-de8a-11e8-9a6c-798789490087.PNG">

**Comment by dgriffen** [comment]
There seems to be an issue with linking when there are exported members in a previous repl line (e.g. pub struct Foo;). The linker error you get is similar to the following:

```
error: linking with `link.exe` failed: exit code: 1104
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Preview\\VC\\Tools\\MSVC\\14.20.27404\\bin\\HostX64\\x64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LIBPATH:C:\\Users\\dagriffe\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "C:\\Users\\dagriffe\\AppData\\Local\\Temp\\.tmpf5G603\\target\\debug\\deps\\user_code_2.3ro1qz0yvfomqxre.rcgu.o" "/OUT:C:\\Users\\dagriffe\\AppData\\Local\\Temp\\.tmpf5G603\\target\\debug\\deps\\user_code_2.dll" "/DEF:C:\\Users\\dagriffe\\AppData\\Local\\Temp\\rustcWgCiN4\\lib.def" "C:\\Users\\dagriffe\\AppData\\Local\\Temp\\.tmpf5G603\\target\\debug\\deps\\user_code_2.21cvc6zuqz0lml37.rcgu.o" "/OPT:REF,ICF" "/DEBUG" "/NATVIS:C:\\Users\\dagriffe\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\intrinsic.natvis" "/NATVIS:C:\\Users\\dagriffe\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\dagriffe\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/LIBPATH:C:\\Users\\dagriffe\\AppData\\Local\\Temp\\.tmpf5G603\\target\\debug\\deps" "/LIBPATH:C:\\Users\\dagriffe\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "/LIBPATH:C:\\Users\\dagriffe\\AppData\\Local\\Temp\\.tmpf5G603\\target\\debug\\deps" "evcxr_internal_runtime.dll.lib" "/LIBPATH:C:\\Users\\dagriffe\\.rustup\\toolchains\\nightly-x86_64-pc-windows-msvc\\lib\\rustlib\\x86_64-pc-windows-msvc\\lib" "std-1f700141642f64de.dll.lib" "C:\\Users\\dagriffe\\AppData\\Local\\Temp\\rustcWgCiN4\\libcompiler_builtins-57c192cd94dba2eb.rlib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "msvcrt.lib" "/DLL" "/IMPLIB:C:\\Users\\dagriffe\\AppData\\Local\\Temp\\.tmpf5G603\\target\\debug\\deps\\user_code_2.dll.lib"
  = note: LINK : fatal error LNK1104: cannot open file 'C:\Users\dagriffe\AppData\Local\Temp\.tmpf5G603\target\debug\deps\user_code_2.dll'
```

**Comment by dgriffen** [comment]
On further investigation this appears to be due to some file locking mechanisms on windows that are preventing the dll from being accessed while evcxr is running. If I close evxcr and run the compilation commands myself the compilation succeeds.

**Comment by happydpc** [comment]
Now the windows can be installed. But when I run the expression, It shows these:

![tim 20190304183932](https://user-images.githubusercontent.com/771776/53728128-e643a280-3eac-11e9-8aec-7ee0f73ca833.png)

Hope it works soon. What you have done is awesome.

**Comment by happydpc** [comment]
Now I fix the problem now with the newest code in the trunk. The zmq library problem can be solved by referring this  post:
https://github.com/erickt/rust-zmq/issues/208#issuecomment-380353800

**Comment by rasa200** [comment]
Hi there!! 

First of all, thanks for the project.
I just tried to install it and worked for me!!! :D (Windows 10 user).
I would like to inform the following:

**Speed**
10 seconds processing the command `let x = 5;`

**Multiple-line**
Functions, other than `pub`, need to be defined in a single line.
